python/src/energuide/extractor.py

```diff
@@ -0,0 +1,51 @@
+import csv
+import typing
+import sys
+import cerberus
+from energuide import reader
+
+
+_SCHEMA = {
+    'EVAL_ID': {'type': 'string', 'required': True},
+    'EVAL_TYPE': {'type': 'string', 'required': True},
+
+    'ENTRYBY': {'type': 'string', 'required': True},
+    'CLIENTPCODE': {'type': 'string', 'required': True},
+    'CLIENTNAME': {'type': 'string', 'required': True},
+    'TELEPHONE': {'type': 'string', 'required': True},
+    'MAIL_PCODE': {'type': 'string', 'required': True},
+    'TAXNUMBER': {'type': 'string', 'required': True},
+
+    'RAW_XML': {'type': 'string', 'required': True}
+    }
+
+DROP_FIELDS = ['ENTRYBY',
+               'CLIENTNAME',
+               'CLIENTPCODE',
+               'TELEPHONE',
+               'MAIL_PCODE',
+               'TAXNUMBER',
+               'RAW_XML']
+
+
+def validated(data: typing.Iterable[reader.InputData], validator) -> typing.Iterator[reader.InputData]:
+    for row in data:
+        if not validator.validate(row):
+            error_keys = ', '.join(validator.errors.keys())
+            raise reader.InvalidInputDataException(f'Validator failed on keys: {error_keys}')
+
+        document = validator.document
+        document['FORWARDSORTATIONAREA'] = document['CLIENTPCODE'][:3]
+        for key in DROP_FIELDS:
+            document.pop(key)
+
+        yield document
+
+
+def extract(csv_input: typing.Iterable[str]) -> typing.Iterator[reader.InputData]:
+    csv.field_size_limit(sys.maxsize)
+
+    validator = cerberus.Validator(_SCHEMA, allow_unknown=True)
+    csv_reader = csv.DictReader(csv_input)
```

Comment:
`csv_input` is declared as just an `Iterable[str]`, but clearly it's meant to be more than that since `DictReader` expects the first `str` to be a header row, etc. You couldn't pass just any iterable of `str` in here and expect it to work, so the typing is insufficient here. (Also note that `str` is an iterable of `str` itself, so you could literally pass `foo` into this function and typing would pass).

I think we're slicing responsibility for this function at the wrong point. Maybe separate it into a separate `_read_csv` function that takes a filepath and returns an `Iterable[Dict]`?

Reply:
It's OK for now to assume the input is a filepath to the whole thing. We'll work out how to switch between input fixtures and a DB connection at some point later.